AnyVal/value classes restrictions

gkossakowski opened the pull request:
This is merge of #1366, #1300 and some more restrictions that I collected from various value classes tickets.

Review by @odersky, @retronym.
//cc @hubertp

odersky committed:
Revised restrictions for value classes and unversal traits

and brought compiler in line with them. One thing we can accept IMO are nested
classes (nested objects are still a problem). In fact, it makes no sense to
exclude nested classes from value classes but not from universal traits. A class
nested in universal trait will becomes a class nested in a value class by
inheritance. Note that the reflection library already contains a universal trait
with a nested class (IndexedSeqLike), so we should accept them if we can.

odersky committed:
Fixed problem in SI-6408

Fixed problem reported in comment, where inner classes of value classe caused a compiler crash.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/675/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1384/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/675/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1384/

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
+          // see https://issues.scala-lang.org/browse/SI-6463
+          case ClassDef(mods, _, _, _) if isValueClass =>
+            implRestriction(stat, s"nested ${ if (mods.isTrait) "trait" else "class" }")
+          case _: Import | _: ClassDef | _: TypeDef | EmptyTree => // OK
```
I believe that (at least) `Block` is missing here. Test with `class X(val x: Any) extends AnyVal { { "".isEmpty; "".isEmpty } }`

retronym replied:
Also `LabelDef`: `class X(val x: Any) extends AnyVal { while (true) ??? }`

retronym replied:
An alternative is to remove `notAllowed("this statement")` below.

retronym replied:
Oh wait, we're trying to prohibit these: "A class or trait C is ephemeral if the following holds: ... C may not have initialization statements". 

I blame it on the hour...

gkossakowski replied:
Yes, I believe that ephemeral class/trait can only have defs in their `Template` tree node so we are good with this set of conditions.

Does it make sense?

retronym replied:
Yep, makes perfect sense.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
+            if (stat.symbol.isAuxiliaryConstructor)
+              notAllowed("secondary constructor")
+            else if (isValueClass && (name == nme.equals_ || name == nme.hashCode_))
+              notAllowed(s"redefinition of $name method")
```
I don't remember the reason for these. Is the list complete? What about `##` and `==`?

retronym replied:
That's right, it's a way to ensure these sort of rewritings are safe:

https://github.com/scala/scala/blob/master/src/compiler/scala/tools/nsc/transform/PostErasure.scala#L48

I can't find the one for `hashCode` right now.

This code should document the reasons for this restriction.

gkossakowski replied:
It's criterion number 5 from SIP:

```
C may not define equals or hashCode methods.
```

Good question about `##` and `==`. I thought that those are treated specially only for case classes, no?

harrah replied:
`##` and `==` are final, so they are handled by other checks.  I think equals/hashCode should be disallowed for universal traits as well, though.  I think that is out of scope for this pull request, though.

retronym replied:
Overloading of these methods can also cause problems, if the optimizations are based on name rather than symbol. But it looks like that is handled correctly now: https://github.com/scala/scala/blob/master/src/compiler/scala/tools/nsc/transform/PostErasure.scala#L55

gkossakowski replied:
@retronym: I added mention of this specific criterion in error message. Please have another look.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
+              notAllowed("secondary constructor")
+            else if (isValueClass && (name == nme.equals_ || name == nme.hashCode_))
+              notAllowed(s"redefinition of $name method")
+            else if (stat.symbol != null && (stat.symbol hasFlag PARAMACCESSOR))
```
I would suggest using one of the Flags helper methods. In this case case it is isParamAccessor.

gkossakowski replied:
I'll fix it. Thanks.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
+      object checkEphemeralDeep extends Traverser {
+          override def traverse(tree: Tree): Unit = {
+            tree match {
+              case _: ModuleDef if isValueClass =>
```
Shouldn't isValueClass guard be lifted?

gkossakowski replied:
Ah yes. Nested objects are disallowed both in universal traits and value classes. Nested classes are only disallowed in value classes.

I'll fix it. Thanks.

gkossakowski commented:
All: I incorporated all of your feedback. Have another look, please.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
+                implRestriction(tree, "nested object")
+              //see https://issues.scala-lang.org/browse/SI-6444
+              // see https://issues.scala-lang.org/browse/SI-6463
+              case _: ClassDef if isValueClass =>
```
What I meant by saying "to lift isValueClass guard" is that it was included in every case of this Traverser

hubertp replied:
And I guess it should be there, just put it before you match on tree.

gkossakowski replied:
Yes, you are right. Sorry, I got confused what is allowed and what is not in universal traits.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
+      // value classes. This code can be cleaned up a lot once implementation
+      // restrictions are addressed.
+      val isValueClass = !clazz.isTrait
+      def where = if (clazz.isTrait) "universal trait extending from class Any" else "value class"
```
You already have negation of the condition in isValueClass value.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1395,7 +1459,7 @@ trait Typers extends Modes with Adaptations with Tags {
           (if (clazz.owner.isTerm) "local class" else "member of another class"))
       if (!clazz.isPrimitiveValueClass) {
-        clazz.info.decls.toList.filter(acc => acc.isMethod && (acc hasFlag PARAMACCESSOR)) match {
+        clazz.info.decls.toList.filter(acc => acc.isMethod && acc.isParamAccessor) match {
```
shouldn't acc.isMethod here be dropped as param accessor will always be a method?

gkossakowski replied:
Good question. This part of code was written by @odersky so I don't know what was the reasoning here. Strictly speaking `isParamAccessor` looks like this:

```
  def isParamAccessor    = hasFlag(PARAMACCESSOR)
```

so the contract we are talking about is not enforced. I hope @odersky can comment on this.

hubertp replied:
Looking at the usage of isParamAccessor in Symbols it seems that my assumption was wrong. So leave it as it is.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/683/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1392/

gkossakowski committed:
Put more implementation restrictions on value classes.

Nested objects, classes and lazy vals are disallowed at any
nesting level in value classes; e.g. lazy vals local to a
method defined in a value class. There are still allowed in
universal traits.

This is a temporary, implementation restriction that is planned
to be addressed in future releases of Scala. Error messages has
been updated to communicate that intent.

Moved tests for SI-5582 and SI-6408 to pending folder. They have
to stay there until implementation restrictions are addressed.

Closes SI-6408 and SI-6432.

Review by @odersky, @harrah and @adriaanm.

gkossakowski commented:
@hubertp: Can you have another look? I hope I got everything right this time.

hubertp commented:
lgtm

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1392/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/683/

gkossakowski commented:
PLS REBUILD ALL

gkossakowski commented:
@retronym: do you want to have another look before I merge? lgtm from hubert is enough for me.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/685/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1394/

retronym commented:
lgtm, too.

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/685/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1394/

gkossakowski commented:
Cool. Thanks a lot!
